### ISSUE TITLE
"current" folder not available in dep4.3.1

### ISSUE THREAD

**Issue by pquerner** [sourcebroker/deployer-extended#7]
https://github.com/sourcebroker/deployer-extended/blob/44ac8515cf21dcfd7c03c42544327298c5c6a870/deployer/buffer/task/buffer:start.php#L62

You reference "current" as the folder where the new files are. This fails in a deployer 4.3.1 installation, because the folder is called "release".

**Comment by kszymukowicz** [comment]
Problem is because you are doing first deploy and "current" symlink does not exist yet. After first successful deploy there will be "current" symlink created.

Anyway its a bug and I will fix that! Tnx for informing.

A TIP for now. 
To have first working deploy you can disable some not required tasks, like for example buffer:start. Just make empty task in deploy.php
 
```
task('buffer:start', function () {});
```

**Comment by pquerner** [comment]
Once the first deploy completed successfull I commented out the overwritten task, then this happend:

```
Can not find file to overwrite or the file is empty. File that was read is: /var/www/deployertest.local.typo3.org/release/index.php
```

Stack
```
dep deploy dev
✔ Executing task deploy:check_lock
✔ Executing task deploy:check_composer_install
✔ Executing task deploy:prepare
✔ Executing task deploy:lock
✔ Executing task deploy:release
✔ Executing task deploy:update_code
✔ Executing task deploy:shared
✔ Executing task deploy:writable
✔ Executing task deploy:vendors
✔ Executing task deploy:clear_paths
✔ Executing task db:backup
➤ Executing task buffer:start
✔ Executing task deploy:failed
✔ Executing task deploy:unlock

In buffer:start.php line 124:
```

Full verbose

```
➤ Executing task deploy:check_lock
• done on [dev]
✔ Ok [0ms]
➤ Executing task deploy:check_composer_install
[localhost] > : which composer
> /usr/local/bin/composer
[localhost] > : which php
> /usr/local/bin/php
[localhost] > : /usr/local/bin/php /usr/local/bin/composer --ignore-platform-reqs install --dry-run 2>&1
> Loading composer repositories with package information
> Installing dependencies (including require-dev) from lock file
> Nothing to install or update
> Package react/socket-client is abandoned, you should avoid using it. Use react/socket instead.
• done on [dev]
✔ Ok [184ms]
➤ Executing task deploy:prepare
[dev] > echo $0
[dev] < bash
[dev] > if [ ! -d /var/www/deployertest.local.typo3.org ]; then mkdir -p /var/www/deployertest.local.typo3.org; fi
[dev] > if [ ! -L /var/www/deployertest.local.typo3.org/current ] && [ -d /var/www/deployertest.local.typo3.org/current ]; then echo true; fi
[dev] > cd /var/www/deployertest.local.typo3.org && if [ ! -d .dep ]; then mkdir .dep; fi
[dev] > cd /var/www/deployertest.local.typo3.org && if [ ! -d releases ]; then mkdir releases; fi
[dev] > cd /var/www/deployertest.local.typo3.org && if [ ! -d shared ]; then mkdir shared; fi
• done on [dev]
✔ Ok [207ms]
➤ Executing task deploy:lock
[dev] > if [ -f /var/www/deployertest.local.typo3.org/.dep/deploy.lock ]; then echo 'true'; fi
[dev] > touch /var/www/deployertest.local.typo3.org/.dep/deploy.lock
• done on [dev]
✔ Ok [74ms]
➤ Executing task deploy:release
[dev] > cd /var/www/deployertest.local.typo3.org && (if [ -h release ]; then echo 'true'; fi)
[dev] < true
[dev] > cd /var/www/deployertest.local.typo3.org && (rm -rf "$(readlink release)")
[dev] > cd /var/www/deployertest.local.typo3.org && (rm release)
[dev] > cd /var/www/deployertest.local.typo3.org && ([ -d releases ] && [ "$(ls -A releases)" ] && echo "true" || echo "false")
[dev] < true
[dev] > cd /var/www/deployertest.local.typo3.org && (cd releases && ls -t -1 -d */)
[dev] < 1/
[dev] > cd /var/www/deployertest.local.typo3.org && (if [ -f .dep/releases ]; then echo "true"; fi)
[dev] < true
[dev] > cd /var/www/deployertest.local.typo3.org && (tail -n 15 .dep/releases)
[dev] < 20180606120224,1
[dev] < 20180606120456,1
[dev] < 20180606120853,1
[dev] < 20180606120950,1
[dev] < 20180606121212,1
[dev] < 20180606121403,1
[dev] < 20180606121743,1
[dev] < 20180606122832,1
[dev] < 20180606123144,1
[dev] < 20180606123826,1
[dev] < 20180606123907,1
[dev] < 20180606124012,1
[dev] < 20180606124238,1
[dev] < 20180606124338,2
[dev] < 20180606124525,2
[dev] > cd /var/www/deployertest.local.typo3.org && (if [ -d /var/www/deployertest.local.typo3.org/releases/2 ]; then echo 'true'; fi)
[dev] > cd /var/www/deployertest.local.typo3.org && (date +"%Y%m%d%H%M%S")
[dev] < 20180606124610
[dev] > cd /var/www/deployertest.local.typo3.org && (echo '20180606124610,2' >> .dep/releases)
[dev] > cd /var/www/deployertest.local.typo3.org && (mkdir /var/www/deployertest.local.typo3.org/releases/2)
[dev] > cd /var/www/deployertest.local.typo3.org && (if [[ "$(man ln)" =~ "--relative" ]]; then echo "true"; fi)
[dev] < true
[dev] > cd /var/www/deployertest.local.typo3.org && (ln -nfs --relative /var/www/deployertest.local.typo3.org/releases/2 /var/www/deployertest.local.typo3.org/release)
• done on [dev]
✔ Ok [485ms]
➤ Executing task deploy:update_code
[dev] > which git
[dev] < /usr/bin/git
[dev] > /usr/bin/git version
[dev] < git version 2.7.4
[dev] > if [ -h /var/www/deployertest.local.typo3.org/release ]; then echo 'true'; fi
[dev] < true
[dev] > readlink /var/www/deployertest.local.typo3.org/release
[dev] < releases/2
[dev] > /usr/bin/git clone -b CUD_17004-208  --recursive -q ssh://xxx/typo3-cms.git /var/www/deployertest.local.typo3.org/releases/2 2>&1
• done on [dev]
✔ Ok [392ms]
➤ Executing task deploy:shared
[dev] > if [ -d /var/www/deployertest.local.typo3.org/shared/fileadmin ]; then echo 'true'; fi
[dev] < true
[dev] > rm -rf /var/www/deployertest.local.typo3.org/releases/2/fileadmin
[dev] > mkdir -p `dirname /var/www/deployertest.local.typo3.org/releases/2/fileadmin`
[dev] > ln -nfs --relative /var/www/deployertest.local.typo3.org/shared/fileadmin /var/www/deployertest.local.typo3.org/releases/2/fileadmin
[dev] > if [ -d /var/www/deployertest.local.typo3.org/shared/uploads ]; then echo 'true'; fi
[dev] < true
[dev] > rm -rf /var/www/deployertest.local.typo3.org/releases/2/uploads
[dev] > mkdir -p `dirname /var/www/deployertest.local.typo3.org/releases/2/uploads`
[dev] > ln -nfs --relative /var/www/deployertest.local.typo3.org/shared/uploads /var/www/deployertest.local.typo3.org/releases/2/uploads
[dev] > if [ -d /var/www/deployertest.local.typo3.org/shared/typo3temp/assets/_processed_ ]; then echo 'true'; fi
[dev] < true
[dev] > rm -rf /var/www/deployertest.local.typo3.org/releases/2/typo3temp/assets/_processed_
[dev] > mkdir -p `dirname /var/www/deployertest.local.typo3.org/releases/2/typo3temp/assets/_processed_`
[dev] > ln -nfs --relative /var/www/deployertest.local.typo3.org/shared/typo3temp/assets/_processed_ /var/www/deployertest.local.typo3.org/releases/2/typo3temp/assets/_processed_
[dev] > if [ -d /var/www/deployertest.local.typo3.org/shared/typo3temp/assets/images ]; then echo 'true'; fi
[dev] < true
[dev] > rm -rf /var/www/deployertest.local.typo3.org/releases/2/typo3temp/assets/images
[dev] > mkdir -p `dirname /var/www/deployertest.local.typo3.org/releases/2/typo3temp/assets/images`
[dev] > ln -nfs --relative /var/www/deployertest.local.typo3.org/shared/typo3temp/assets/images /var/www/deployertest.local.typo3.org/releases/2/typo3temp/assets/images
[dev] > if [ -d /var/www/deployertest.local.typo3.org/shared/typo3temp/var/logs ]; then echo 'true'; fi
[dev] < true
[dev] > rm -rf /var/www/deployertest.local.typo3.org/releases/2/typo3temp/var/logs
[dev] > mkdir -p `dirname /var/www/deployertest.local.typo3.org/releases/2/typo3temp/var/logs`
[dev] > ln -nfs --relative /var/www/deployertest.local.typo3.org/shared/typo3temp/var/logs /var/www/deployertest.local.typo3.org/releases/2/typo3temp/var/logs
[dev] > if [ -f $(echo /var/www/deployertest.local.typo3.org/releases/2/.env) ]; then rm -rf /var/www/deployertest.local.typo3.org/releases/2/.env; fi
[dev] > if [ ! -d $(echo /var/www/deployertest.local.typo3.org/releases/2/.) ]; then mkdir -p /var/www/deployertest.local.typo3.org/releases/2/.;fi
[dev] > mkdir -p /var/www/deployertest.local.typo3.org/shared/.
[dev] > touch /var/www/deployertest.local.typo3.org/shared/.env
[dev] > ln -nfs --relative /var/www/deployertest.local.typo3.org/shared/.env /var/www/deployertest.local.typo3.org/releases/2/.env
• done on [dev]
✔ Ok [883ms]
➤ Executing task deploy:writable
[dev] > ps axo user,comm | grep -E '[a]pache|[h]ttpd|[_]www|[w]ww-data|[n]ginx' | grep -v root | head -1 | cut -d\  -f1
[dev] < vagrant
[dev] > cd /var/www/deployertest.local.typo3.org/releases/2 && (mkdir -p typo3conf typo3temp uploads fileadmin)
[dev] > cd /var/www/deployertest.local.typo3.org/releases/2 && ( chown -RL vagrant typo3conf typo3temp uploads fileadmin)
• done on [dev]
✔ Ok [111ms]
➤ Executing task deploy:vendors
[dev] > if hash composer 2>/dev/null; then echo 'true'; fi
[dev] < true
[dev] > which composer
[dev] < /usr/local/bin/composer
[dev] > which php
[dev] < /usr/bin/php
[dev] > cd /var/www/deployertest.local.typo3.org/releases/2 &&  /usr/bin/php /usr/local/bin/composer install --verbose --prefer-dist --no-progress --no-interaction --no-dev --optimize-autoloader
[dev] < Generating  class alias map file
[dev] < Inserting class alias loader into main autoload.php file
• done on [dev]
✔ Ok [7s 505ms]
➤ Executing task deploy:clear_paths
[dev] >  rm -rf /var/www/deployertest.local.typo3.org/releases/2/.git
[dev] >  rm -rf /var/www/deployertest.local.typo3.org/releases/2/.gitignore
[dev] >  rm -rf /var/www/deployertest.local.typo3.org/releases/2/composer.json
[dev] >  rm -rf /var/www/deployertest.local.typo3.org/releases/2/composer.lock
[dev] >  rm -rf /var/www/deployertest.local.typo3.org/releases/2/composer.phar
[dev] >  rm -rf /var/www/deployertest.local.typo3.org/releases/2/.gitattributes
[dev] >  rm -rf /var/www/deployertest.local.typo3.org/releases/2/.env.dist
[dev] >  rm -rf /var/www/deployertest.local.typo3.org/releases/2/dynamicReturnTypeMeta.json
• done on [dev]
✔ Ok [284ms]
➤ Executing task db:backup
• done on [dev]
✔ Ok [0ms]
➤ Executing task buffer:start
[dev] > if [ -e /var/www/deployertest.local.typo3.org/release/ ]; then echo 'true'; fi
[dev] < true
[dev] > [ -e /var/www/deployertest.local.typo3.org/release/b83f2f96cdedfe6de076168ec67d1178 ] || mkdir -p /var/www/deployertest.local.typo3.org/release/b83f2f96cdedfe6de076168ec67d1178
[dev] > [ -e /var/www/deployertest.local.typo3.org/release//current ] || touch /var/www/deployertest.local.typo3.org/current/.flag.oldinstance
[dev] > cd /var/www/deployertest.local.typo3.org/releases/2 && [ -f index.php ] && cat index.php || echo ""
[dev] < 
➤ Executing task deploy:failed
• done on [dev]
✔ Ok [0ms]
➤ Executing task deploy:unlock
[dev] > rm -f /var/www/deployertest.local.typo3.org/.dep/deploy.lock
• done on [dev]
✔ Ok [37ms]

```

(I redacted a few things)

**Comment by kszymukowicz** [comment]
Seems like the file index.php is empty. 
How do you install TYPO3 ? 
Do you see any content in index.php ?

**Comment by kszymukowicz** [comment]
Please use this composer.json as example to install TYPO3.
```
{
  "name": "v/base",
  "license": "",
  "config": {
    "process-timeout": 99999,
    "platform": {
      "php": "7.1"
    }
  },
  "repositories": [
    {
      "type": "composer",
      "url": "https://composer.typo3.org/"
    }
  ],
  "require": {
    "typo3/cms-core": "~8.7.0",

    "typo3/cms-belog": "~8.7.0",
    "typo3/cms-beuser": "~8.7.0",
    "typo3/cms-context-help": "~8.7.0",
    "typo3/cms-cshmanual": "~8.7.0",
    "typo3/cms-fluid-styled-content": "~8.7.0",
    "typo3/cms-filelist": "~8.7.0",
    "typo3/cms-impexp": "~8.7.0",
    "typo3/cms-info": "~8.7.0",
    "typo3/cms-info-pagetsconfig": "~8.7.0",
    "typo3/cms-lowlevel": "~8.7.0",
    "typo3/cms-reports": "~8.7.0",
    "typo3/cms-rsaauth": "~8.7.0",
    "typo3/cms-rte-ckeditor": "~8.7.0",
    "typo3/cms-recycler": "~8.7.0",
    "typo3/cms-setup": "~8.7.0",
    "typo3/cms-sys-note": "~8.7.0",
    "typo3/cms-tstemplate": "~8.7.0",
    "typo3/cms-viewpage": "~8.7.0",

    "jigal/t3adminer": "~9.0.0",
    "fluidtypo3/vhs": "~4.3.0",
    "dmitryd/typo3-realurl": "~2.3.0",

    "sourcebroker/deployer-extended-typo3": "^6.0.0"
  },
  "extra": {
    "typo3/cms": {
      "cms-package-dir": "{$vendor-dir}/typo3/cms",
      "web-dir": "."
    },
    "helhum/typo3-console": {
      "install-extension-dummy": false
    },
    "helhum/dotenv-connector": {
      "env-dir": "",
      "allow-overrides": true,
      "cache-in-dev-mode": true
    }
  },
  "scripts": {
    "packages-states": [
      "export $(cat .env | xargs) && ./vendor/bin/typo3cms install:generatepackagestates"
    ],
    "folder-structure": [
      "export $(cat .env | xargs) && ./vendor/bin/typo3cms install:fixfolderstructure"
    ],
    "ext-setup": [
      "export $(cat .env | xargs) && ./vendor/bin/typo3cms install:extensionsetupifpossible"
    ],
    "post-autoload-dump": [
      "@packages-states",
      "@folder-structure",
      "@ext-setup"
    ]
  }
}

```

**Comment by pquerner** [comment]
```
ls -lah
total 36K
drwxrwxr-x  7 vagrant vagrant 4.0K Jun  6 12:46 .
drwxrwxr-x  4 vagrant vagrant 4.0K Jun  6 12:46 ..
drwxrwxr-x  2 vagrant vagrant 4.0K Jun  6 12:46 b83f2f96cdedfe6de076168ec67d1178
drwxrwxr-x  2 vagrant vagrant 4.0K Jun  6 12:46 typo3conf
drwxrwxr-x  4 vagrant vagrant 4.0K Jun  6 12:46 typo3temp
drwxrwxr-x 24 vagrant vagrant 4.0K Jun  6 12:46 vendor
drwxrwxr-x  3 vagrant vagrant 4.0K Jun  6 12:46 web
-rw-rw-r--  1 vagrant vagrant 1.9K Jun  6 12:46 deploy.php
lrwxrwxrwx  1 vagrant vagrant   17 Jun  6 12:46 .env -> ../../shared/.env
-rw-rw-r--  1 vagrant vagrant  132 Jun  6 12:46 .env.sample
lrwxrwxrwx  1 vagrant vagrant   22 Jun  6 12:46 fileadmin -> ../../shared/fileadmin
lrwxrwxrwx  1 vagrant vagrant   20 Jun  6 12:46 uploads -> ../../shared/uploads
/v/w/d/release $ pwd
/var/www/deployertest.local.typo3.org/release
```

As you can see there is no index.php, because that file is in ```web``` dir.

```
ls -lah web
total 12K
drwxrwxr-x 3 vagrant vagrant 4.0K Jun  6 12:46 .
drwxrwxr-x 7 vagrant vagrant 4.0K Jun  6 12:46 ..
drwxrwxr-x 3 vagrant vagrant 4.0K Jun  6 12:46 typo3conf
lrwxrwxrwx 1 vagrant vagrant   29 Jun  6 12:46 index.php -> ../vendor/typo3/cms/index.php
lrwxrwxrwx 1 vagrant vagrant   25 Jun  6 12:46 typo3 -> ../vendor/typo3/cms/typo3
```

TYPO3 CMS is installed via composer.
composer.json
```
{
  "repositories": [
    {
      "type": "composer",
      "url": "https://composer.typo3.org/"
    },
    {
      "type": "package",
      "package": {
        "name": "typo3/extension-builder",
        "version": "dev-master",
        "type": "typo3-cms-extension",
        "source": {
          "url": "https://github.com/FriendsOfTYPO3/extension_builder.git",
          "type": "git",
          "reference": "master"
        }
      }
    },
  ],
  "require": {
    "typo3/cms": "^8.7",
    "bk2k/bootstrap-package": "^9.1",
    "dmitryd/typo3-realurl": "^2.3",
    "in2code/powermail_cond": "^4.1",
    "fluidtypo3/vhs": "^4.3",
    "sourcebroker/deployer-extended-typo3": "^6.0",
    "deployer/deployer": "4.3.1",
    "sourcebroker/deployer-bulk-tasks": "^2.2",
    "sourcebroker/deployer-extended": "^8.0"
  },
  "require-dev": {
    "typo3/extension-builder": "dev-master"
  },
  "extra": {
    "typo3/cms": {
      "cms-package-dir": "{$vendor-dir}/typo3/cms",
      "web-dir": "web"
    }
  },
  "autoload-dev": {
    "psr-4": {
      "EBT\\ExtensionBuilder\\": "web/typo3conf/ext/extension_builder/Classes/",
      "PhpParser\\": "web/typo3conf/ext/extension_builder/Resources/Private/PHP/PHP-Parser/lib/PhpParser/"
    }
  }
}

```
(again, I removed some private repos from this file)

Current deploy.php

```
<?php
namespace Deployer;

//require 'recipe/typo3.php';
require_once(__DIR__ . '/vendor/sourcebroker/deployer-loader/autoload.php');
new \SourceBroker\DeployerExtendedTypo3\Loader();

// Project name
set('application', 'xxx');
set('ssh_multiplexing', true);
set('writable_mode', 'chown');
set('branch', 'CUD_17004-208');

set('db_databases',
    [
        'database_default' => [
            'ignore_tables_out' => [
                'caching_*'
            ],
            (new \SourceBroker\DeployerExtendedDatabase\Driver\EnvDriver())->getDatabaseConfig()
         ],
    ]
);

// Project repository
set('repository', 'ssh://git@xxx/typo3-cms.git');

// [Optional] Allocate tty for git clone. Default value is false.
set('git_tty', true);

// Shared files/dirs between deploys
//add('shared_files', []);
//add('shared_dirs', []);

// Writable dirs by web server
//add('writable_dirs', []);
set('allow_anonymous_stats', false);

// Hosts


//Deployer 6+
//host('192.168.144.120')
//    ->user('vagrant')
//    ->forwardAgent(true)
//    ->multiplexing(true)
//    ->set('deploy_path', '/var/www/deployertest.local.typo3.org');

//Deployer 4.3.1
server('dev', '192.168.144.120')
    ->user('vagrant')
    ->forwardAgent(true)
    ->set('deploy_path', '/var/www/deployertest.local.typo3.org')
    ->set('public_urls', ['http://deployertest.local.typo3.org/release/web']);

// Tasks

task('build', function () {
    run('cd {{release_path}}');
    run('composer install');
});

//Overwrite these too, because they cause errors
//task('deploy:lock', function () {
//
//});
//task('deploy:unlock', function () {
//
//});
//@see https://github.com/sourcebroker/deployer-extended/issues/7
//task('buffer:start', function () {});
task('db:backup', function () {});
task('typo3cms:database:updateschema', function () {});

// [Optional] if deploy fails automatically unlock.
after('deploy:failed', 'deploy:unlock');


```

**Comment by kszymukowicz** [comment]
I can see that your composer config installs into /web/.
I do not use it and its not tested.

**Comment by kszymukowicz** [comment]
Actually buffer:start support storing in any folder. Put in your deploy.php

```
set('buffer_config', [
        'index.php' => [
            'entrypoint_filename' => 'web/index.php', // frontend
        ]
        'typo3/index.php' => [
            'entrypoint_filename' => 'web/typo3/index.php', // backend
        ],
        'typo3/cli_dispatch.phpsh' => [
            'entrypoint_filename' => 'web/typo3/cli_dispatch.phpsh', // cli
        ]
    ]
);
```

Anyway I am not sure how well tested with /web/ are other parts of deployer-extended-typo3.

**Comment by pquerner** [comment]
/web/ is in the default composer.json, thats why I used it. That way there is no way to expose /vendor/ by accident.

**Comment by kszymukowicz** [comment]
I know. Thats arguable. 
There is lot of places in TYPO3 itself protected with .htaccess "deny all". So why its safe to protect one resource like that and not "vendor" folder.

Sooner or later I will switch for /web/ also but its not so critical for me.

**Comment by pquerner** [comment]
/vendor/ by default doesnt get .htaccess protected and must be done manually. I think this is why they did that.
Might be able to just create a deployer task for this, which adds a .htaccess file to that directory.

**Comment by kszymukowicz** [comment]
Are you sure vendor is not protected by default? Check this default htaccess fro TYPO3.
https://github.com/TYPO3/TYPO3.CMS/blob/master/_.htaccess

**Comment by pquerner** [comment]
Ah right, I forgot about that one. I just searched for a .htaccess file in /vendor/ itself. You are right then.

**Comment by kszymukowicz** [comment]
Issue fixed in 9.0.0.
